### PR TITLE
Support dot syntax in action arg keys

### LINKED ISSUE #7399: amp-bind can't set state for variables inside amp-state
At the moment it's only possible to set the values for global variables via AMP.setState but not for variables declared inside amp-state.

Example:
`AMP.setState(foo=123)` -> is working

`AMP.setState(myAmpState.foo=123) `
where 
```
<amp-state id="myAmpState">
      <script type="application/json">
      {
          "foo": "234"
      }
.....
```
is not working.
Current workaround is setting default variables.

@choumx @ericlindley-g

### PR DESCRIPTION
**Pull request by choumx**
Fixes #7399, partial for #6199.

Adds support for nested keys in action method args, e.g. `AMP.setState(foo.bar.baz='qux')`.

### CODE REVIEW

File: src/service/action-impl.js [hunk 1]
```diff
@@ -399,6 +399,18 @@ export function parseActionMap(s, context) {
...
 
+  // Consumes all subsequent token pairs of the form "<.><ID>" and invokes
+  // `callback` for each <ID> token. Used for variable dereferences.
+  function tokenizeIdDereferences(tokenizer, callback) {
+    for (let peek = tokenizer.peek();
+        peek.type == TokenType.SEPARATOR && peek.value == '.';
+        peek = tokenizer.peek()) {
+      tokenizer.next(); // Skip '.'.
+      const token = assertToken(tokenizer.next(false), [TokenType.ID]);
+      callback(token);
```
Comment (jridgewell) [inline src/service/action-impl.js]: Pushing into an (optionally passed in) array will might be better?

Comment (choumx) [inline src/service/action-impl.js]: Both callers would have to iterate through the array which is wasted work. From an API POV, this is similar to `forEach`, `map` or other APIs that take a callback.

Comment (jridgewell) [inline src/service/action-impl.js]: > Both callers would have to iterate through the array which is wasted work

How so?

```js
// first caller
const dereferences = tokenizeIdDereferences(toks);
const argKey = tok.value + dereferences.join('.');

// second caller
tokenizeIdDereferences(toks, argValueTokens);
```

Definitely not necessary.

Comment (choumx) [inline src/service/action-impl.js]: `join` iterates through the array, no?

Why is returning an array here better? TBH I don't have a strong opinion so I'll change it if you insist. 😃

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -115,6 +116,30 @@ export class Bind {
   }
 
   /**
+   * Same as `setState()` except this method supports keys nested with '.'.
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: Does it make sense to have two APIs? Is it not possible to contain everything in one?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 3]
```diff
@@ -115,6 +116,30 @@ export class Bind {
...
+  setStateWithNestedKeys(state) {
+    const unnestedState = map();
+
+    Object.keys(state).forEach(key => {
+      let value = state[key];
+      const fragments = key.split('.');
+      // Create a wrapping object for each '.' in key, starting from inside.
+      // Stop before the first fragment, which is the top-most key.
+      for (let i = fragments.length - 1; i >= 1; i--) {
+        const wrapper = map();
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: Doesn't this overwrite the same keys in the same input  map?

E.g. base has two keys `p.a` and `p.b`. It'd seem that this algorithm would first create object `{p: {a: 1}}` and then overwrite it with `{p: {b: 2}}`? In other words, shouldn't this line be something like:
```
const wrapper = untestedState[fragments[i]] || map();
```
?

And further, should we continue overriding the internal state the same way? Or merge?

**Comment by choumx** [comment]
I'm considering dropping this PR in favor of a more powerful solution, as suggested by Dima, that allows `amp-bind` to parse the action expression within `AMP.setState`.

This would make the syntax of `AMP.setState` consistent with `amp-bind` expressions and support more use cases requested by @kul3r4.

**Comment by choumx** [comment]
Planning to close this in favor of #7573.

**Pull request closed without merging**